metrics/fidelity/fidelity.py

```diff
@@ -0,0 +1,145 @@
+import logging
+
+import numpy as np
+
+import metrics.fidelity.utility as utility
+import datasets
+
+
+logger = logging.getLogger(__name__)
+
+_CITATION = """
+citation is under construction 
+@misc{carton2020evaluating,
+      title={Evaluating and Characterizing Human Rationales}, 
+      author={Samuel Carton and Anirudh Rathore and Chenhao Tan},
+      year={2020},
+      eprint={2010.04736},
+      archivePrefix={arXiv},
+      primaryClass={cs.CL}
+}
+"""
+
+_DESCRIPTION = """
+description is under construction
+sufficiency(x, y_hat, alpha) = 1 - max(0, P(y_hat/x) - P(y_hat/x, alpha))
+comprehensiveness(x, y_hat, alpha) = max(0, P(y_hat/x) - P(y_hat/x, alpha))
+null difference = max(0, P(y_hat/x) - P(y_hat/x, 0))
+"""
+
+_KWARGS_DESCRIPTION = """
+Args:
+	`predictions`: (numpy array) y_hat
+	`prob_y_hat`: (numpy array) P(y_hat/x)
+	`prob_y_hat_alpha`: (numpy array) P(y_hat/x, alpha)
+	`null_difference`: (numpy array) max(0, P(y_hat/x) - P(y_hat/x, 0))
+	`model`: (pytorch nn Module) model to compute predictions
+	`tokenizer`: data tokenizer
+	`input_ids`: (torch.tensor) tensor of input ids
+	`alpha`: (list) 0s and 1s corresponding to the presence of a token in the rationale
+	`attention_masks`: [Optional] (torch.tensor) tensor of attention masks
+	`fidelity_type`: [Default="sufficiency"] (str) sufficiency or comprehensiveness
+	`clip`: [Default=True] (bool) true for clipping the fidelity values between 0 and 1
+	`normalization`: [Default=True] (bool) true for calibrating 
+Returns:
+	`fidelity`: 
+"""
+
+
+class Fidelity(datasets.Metric):
+    def _info(self):
+        return datasets.MetricInfo(
+            description=_DESCRIPTION,
+            citation=_CITATION,
+            inputs_description=_KWARGS_DESCRIPTION,
+            features=datasets.Features(
+                {
+                    "predictions": datasets.Value("float", id="sequence"),
+                    "prob_y_hat": datasets.Value("float", id="sequence"),
+                    "prob_y_hat_alpha": datasets.Value("float", id="sequence"),
+                    "null_difference": datasets.Value("float", id="sequence"),
+                    "model": datasets.Value("float", id="sequence"),
+                    "tokenizer": datasets.Value("float", id="sequence"),
+                    "mode": datasets.Value("string", id="sequence"),
+                    "normalization": datasets.Value("bool", id="sequence"),
+                }
+            ),
+        )
+
+    def _compute(
+        self,
+        predictions=None,
+        prob_y_hat=None,
+        prob_y_hat_alpha=None,
+        null_difference=None,
+        model=None,
+        tokenizer=None,
+        input_ids=None,
+        alpha=None,
+        attention_masks=None,
+        fidelity_type="sufficiency",
+        clip=True,
+        normalization=True,
+        # reduction='mean' # discuss with sam
+    ):
+
+        if normalization:
```

Comment:
Add a comment here explaining that we need either rationale predictions or a model and a rationale with which to make predictions

Reply:
`if normalization:
            if (
                (predictions is None)
                and (prob_y_hat is None or prob_y_hat_alpha is None or null_difference is None)
                and (model is None or input_ids is None or alpha is None)
            ):
                return "Please provide either predictions or model and inputs to compute predictions"
        else:
            if (
                (predictions is None)
                and (prob_y_hat is None or prob_y_hat_alpha is None)
                and (model is None or input_ids is None or alpha is None)
            ):
                return "Please provide either predictions or model and inputs to compute predictions"`
I do a check for this here and return with an error if something is missing